arch/risc-v/src/common/riscv_addrenv.c

```diff
@@ -0,0 +1,828 @@
+ *                         address environment
+ *   up_addrenv_heapsize - Returns the size of the initial heap allocation.
+ *   up_addrenv_select   - Instantiate an address environment
+ *   up_addrenv_restore  - Restore an address environment
+ *   up_addrenv_clone    - Copy an address environment from one location to
+ *                        another.
+ *
+ * Higher-level interfaces used by the tasking logic.  These interfaces are
+ * used by the functions in sched/ and all operate on the thread which whose
+ * group been assigned an address environment by up_addrenv_clone().
+ *
+ *   up_addrenv_attach   - Clone the address environment assigned to one TCB
+ *                         to another.  This operation is done when a pthread
+ *                         is created that share's the same address
+ *                         environment.
+ *   up_addrenv_detach   - Release the threads reference to an address
+ *                         environment when a task/thread exits.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <errno.h>
+#include <assert.h>
+#include <debug.h>
+
+#include <nuttx/addrenv.h>
+#include <nuttx/arch.h>
+#include <nuttx/irq.h>
+#include <nuttx/pgalloc.h>
+
+#include <arch/barriers.h>
+
+#include "pgalloc.h"
+#include "riscv_mmu.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* Only CONFIG_BUILD_KERNEL is supported (i.e. tested) */
+
+#ifndef CONFIG_BUILD_KERNEL
+#  error "This module is intended to be used with CONFIG_BUILD_KERNEL"
+#endif
+
+/* Entries per PGT */
+
+#define ENTRIES_PER_PGT     (RV_MMU_ENTRIES_PER_PGT)
+
+/* Base address for address environment */
+
+#define ADDRENV_VBASE       (CONFIG_ARCH_DATA_VBASE)
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+extern uintptr_t            g_kernel_mappings;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: wipe_page
+ *
+ * Description:
+ *   Wipe a page of physical memory, first mapping it into virtual memory.
+ *
+ * Input Parameters:
+ *   paddr - Physical address of page
+ *
+ ****************************************************************************/
+
+static inline void wipe_page(uintptr_t paddr)
+{
+  uintptr_t vaddr = riscv_pgpool_to_vaddr(paddr);
+  memset((void *)vaddr, 0, MM_PGSIZE);
```

Comment:
do we need to check `vaddr != 0` or this is insured by the caller? I mean maybe adding `DEBUGASSERT(vaddr != 0);` can be reasonable here

Reply:
Ensured by the caller